### ISSUE TITLE
COUNT statement

### ISSUE THREAD

**Issue by ghost** [sockeqwe/sqlbrite-dao#16]
this can be done with the SELECT function, but a high level function can ease the effort of doing this, will do PR if you want something like this

**Comment by sockeqwe** [comment]
I'm not sure how to add that into the SELECT function easily. How would you do that?

**Comment by ghost** [comment]
Okay on the risk of making a joke of my sql knowledge the count function is
done by SELECT COUNT(column names)  FROM TABLE NAME right... implementing
this with the current SELECT function in the Dao class is pretty
straightforward, it's just some sugar coating for the api

**Comment by ghost** [comment]
I think with the current SELECT function it will be like
SELECT("COUNT(*)").FROM...

**Comment by sockeqwe** [comment]
yeah, but actually that would be:

``` java
SELECT("COUNT(*)").FROM() ...
```

and if I will select other things as well:

``` java
SELECT("COUNT(*), column1, colum2").FROM() ...
```

or with AS to name the COUNT column result:

``` java
SELECT("COUNT(*) AS foo, column1, colum2").FROM() ...
```

I'm not sure how this could be expressed?

**Comment by ghost** [comment]
i primarily see COUNT as a function in Dao that just returns the count and not to be used with other columns, so maybe something like

```
public SELECT COUNT(String column){
return SELECT("COUNT("+coumn+")");
}

```

Didn't want to do all that String appending, especially since the other SQL functions in the Dao are so convenient, maybe just me feel free to ignore this if you feel its unnecessary

**Comment by sockeqwe** [comment]
I understand what you mean, but it to me, personally, this SQL grammar seems to be wrong since sql COUNT() is a function like MAX(), MIN(), AVG(), SUM(), etc. and can be used for selecting like any other column and in combination with any arbitrary other function or column name.

I personally prefer correct SQL grammar, even if that means to deal with string concatenation.

**Comment by ghost** [comment]
Hmm one interesting thing maybe if we can chain functions like these to SELECT call, although it does take away the freedom of deciding on the order of the sql functions

**Comment by ghost** [comment]
Something like SELECT(columns...).COUNT().MAX(). FROM()

**Comment by ghost** [comment]
Although not sure if this won't be misleading as to  say that these are not SQL functions but something that the api invented

**Comment by sockeqwe** [comment]
yes, something like that could work, but sometimes the order of selecting columns (or functions matters).

I think more about something like this:
Changing definition of `SELECT(String ... columns)` to `SELECT(CharSequence ... columns)`. Since `CharSequence` is a interface implemented by String we could also provide our own "classes" like `COUNT implements CharSequence` like this 

``` java
class COUNT implements CharSequence {

   private String columnName;
   private String as;

   public COUNT(String columnName){
        this.columnName = columnName;
    }

   public COUNT(String columnName, String as){
       this(columnName);
       this.as = as;
   }

   @Override
   public void toString(){
       if (as != null)
          return "COUNT("+columName+") AS "+as;

       return "COUNT("+columnName+");
   }
}
```

Then you could use it as before:

``` java
SELECT("column1", COUNT("column2"), "column3").FROM() ...
```

**Comment by ghost** [comment]
Yes that's it this is much better it will make the code look less ugly
especially when the number of SQL functions used are too many and also give
the freedom of ordering

**Comment by sockeqwe** [comment]
But before spending more time on that (there are quite a lot functions) I want to know in which direction `SQLBrite` will go (adding first party implementation for SQL grammar)
https://github.com/square/sqlbrite/issues/32

**Comment by sockeqwe** [comment]
There is a much simpler solution for that: Simply add a method to `Dao` base class  that retruns a string, something like this:

``` java
class Dao {
   ...
  protected String COUNT(String columns){
     return "COUNT("+column+")";
}
```

Then in your `MyDao`

``` java
class MyDao extends Dao {
   public Observable<FooCount> getFooCount(){
        return query( SELECT ("col1", COUNT("*"), "col2").FROM("Foo") ...
   }
}
```
